### ISSUE TITLE
Newly Observed Domain Tracking slows down resolving in certain circumstances

### ISSUE THREAD

**Issue by paddg** [PowerDNS/pdns#8836]
- Program: Recursor
 - Issue type: Bug report

### Short description
If I use `new-domain-tracking` in conjunction with `new-domain-lookup` and `forward-zones`, resolving of "newly observed domains" is dog slow. But only if I send the `new-domain-lookup` to an IP address, which is not responding (neither DNS nor ICMP).

### Environment
 - Operating system: SLES15-SP1
 - Software version: PowerDNS Recursor 4.2.1
 - Software source: self compiled

### Steps to reproduce
1. cd into an empty directory
`# mkdir -p nod-test && cd nod-test`
2. Delete possibly existing new-domain-history files and start the Recursor from command line
`# rm -f nod_*; /usr/sbin/pdns_recursor --daemon=no --config-dir=. --local-address=127.0.0.53 --new-domain-tracking=yes --new-domain-history-dir=. --new-domain-lookup=new-domain-lookup --forward-zones=new-domain-lookup=100::`
3. dig it
```
$ dig +tries=1 +time=10 @127.0.0.53 . SOA
...
;; Query time: 1506 msec
...
```
### Expected behaviour
"Query time:" below 100 msec

### Actual behaviour
"Query time:" above 1000 msec

**Comment by rgacogne** [comment]
At first glance it looks like this code calls `directResolve()` before the response to the initial query has been sent, so we will wait for an answer to come back before sending the "real" response. It would be better to delay the sending of the NOD query until after sending the "real" response has been sent if possible, and perhaps we should not wait for the NOD response at all?

**Comment by paddg** [comment]
I think waiting is not necessary because "The result of the DNS lookup will be ignored by the recursor" (doc) anyway.

**Comment by neilcook** [comment]
> I think waiting is not necessary because "The result of the DNS lookup will be ignored by the recursor" (doc) anyway.

This true. I'm not aware of a "fire-and-forget" version of directResolve(), or maybe we could just create one?

**Comment by rgacogne** [comment]
We don't have a "fire-and-forget" version of `directResolve()`. And to be clear we can't have a version that "just" sends some `UDP` packet and returns because we might need to do recursive resolution and wait for responses from intermediate authoritative servers before we can send the final packet. So at the moment I believe the best solution is to delay the sending of the NOD query until after the "real" response has been sent, allowing us to keep using the existing `mthread` without impacting the service latency.

**Comment by neilcook** [comment]
Yep make sense

**Comment by paddg** [comment]
Since we've passed 4.3.0 milestone already, do you have new plans for fixing this?

**Comment by rgacogne** [comment]
Hi! Sorry about the delay. I just opened a PR implementing the work-around discussed above, and we should be able to backport it to 4.3.x so it's shipped in 4.3.1.

**Comment by paddg** [comment]
That's good news :-) Thanks! 
> so it's shipped in 4.3.1

I assume you mean 4.3.2 as 4.3.1 is already out

**Comment by rgacogne** [comment]
Oh, yes!
